HeterogeneousCore/AlpakaCore/README.md

```diff
@@ -0,0 +1,403 @@
+
+## Introduction
+
+This page documents the Alpaka integration within CMSSW. For more information about Alpaka itself see the [Alpaka documentation](https://alpaka.readthedocs.io/en/latest/).
+
+### Compilation model
+
+The code in `Package/SubPackage/{interface,src,plugins,test}/alpaka` is compiled once for each enabled Alpaka backend. The `ALPAKA_ACCELERATOR_NAMESPACE` macro is substituted with a concrete, backend-specific namespace name in order to guarantee different symbol names for all backends, that allows for `cmsRun` to dynamically load any set of the backend libraries.
+
+The source files with `.dev.cc` suffix are compiled with the backend-specific device compiler. The other `.cc` source files are compiled with the host compiler.
+
+The `BuildFile.xml` must contain `<flags ALPAKA_BACKENDS="1"/>` to enable the behavior described above.
+
+## Overall guidelines
+
+* Minimize explicit blocking synchronization calls
+  * Avoid `alpaka::wait()`, non-cached memory buffer allocations
+* If you can, use `global::EDProducer` base class
+  * If you need per-stream storage
+    * For few objects consider using [`edm::StreamCache<T>`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_StreamCacheT) with the global module, or
+    * Use `stream::EDProducer`
+  * If you need to transfer some data back to host, use `stream::SynchronizingEDroducer`
+* All code using `ALPAKA_ACCELERATOR_NAMESPACE` should be placed in `Package/SubPackage/{interface,src,plugins,test}/alpaka` directory
+  * Alpaka-dependent code that uses templates instead of the namespace macro can be placed in `Package/SubPackage/interface` directory
+* All source files (not headers) using Alpaka device code (such as kernel call, functions called by kernels) must have a suffic `.dev.cc`, and be placed in the aforementioned `alpaka` subdirectory
+* Any code that `#include`s a header from the framework or from the `HeterogeneousCore/AlpakaCore`(?) must be separated from the Alpaka device code, and have the usual `.cc` suffix.
+  * A few framework headers are allowed to be used in `.dev.cc` files, e.g. `FWCore/Utilities/interface/Exception.h` or `FWCore/MessageLogger/interface/MessageLogger.h` (?)
+
+## Data formats
+
+Data formats, for both Event and EventSetup, should be placed following their usual rules. The Alpaka-specific conventions are
+* There must be a host-only flavor of the data format that is either idependent of Alpaka, or depends only on Alpaka's Serial backend
+  * The host-only data format must be defined in `Package/SubPackage/interface/` directory
+  * If the data format is to be serialized (with ROOT), it must be serialized in a way that the on-disk format does not depend on Alpaka, i.e. it can be read without Alpaka
+  * For Event data products the ROOT dictionary should be defined in `Package/SubPackage/src/classes{.h,_def.xml}` (**NOTE**: currently they are in `.../src/alpaka/classes_serial{.h,_def.xml}`
```

Comment:
I think it would be better to define the dictionaries for the host-only data format classes in `src/classes_def.xml` instead of `src/alpaka/classes_serial_def.xml`, because it would be more consistent with "host platform data formats are usable by non-Alpaka modules". Technically ROOT takes care of loading the proper shared object of the dictionary, so the impact is close to invisible in practice. At least we would be able to avoid building the shared object for the serial backend.

Reply:
I've been thinking about this as well, and it would also simplify supporting different host back-ends like TBB.
I'm mostly waiting to see if Eric's work with dictionaries can simplify things further, but I'm OK with the move.

Reply:
To be clear: if you prefer to move them already now, I'm fine with that.

Reply:
Since the move is not particularly urgent, how about, to avoid further conflicts with #40285, we don't touch the existing `classes_serial_def.xml` files now, but move them after #40285, and follow the guideline for any other new code that may be PR'd before that?

(although I would be fine with moving them already now as well)

Reply:
I don't think we will be able to review and merge #40285 this week, as we'll be busy with the tutorial until Friday. If you'd like to move them, do go ahead.

My only request is to base the branch with these changes on top of 3cd90b2bce8, which is itself based on `CMSSW_13_0_0_pre4`, in order to try and keep a single branch valid for both `CMSSW_13_0_X` and `CMSSW_13_1_X`.

Reply:
Fair enough. Maybe I'll move them then.